File: 003/003.rb
#!/usr/bin/env ruby -w

=begin

Problem 3 (02 November 2001)

Find the largest prime factor of a composite number.

---

The prime factors of 13195 are 5, 7, 13 and 29.

What is the largest prime factor of the number 600851475143 ?

---

Answer:

6857

Took: 0.020021 secs.

real    0m0.048s
user    0m0.035s
sys     0m0.009s

4/27/2012

=end

t = Time.now

require 'mathn'

NUM = 600_851_475_143

puts NUM.prime_division.flatten.uniq.sort.last

puts "\nTook: #{(Time.now - t)} secs."

File: 003/003.c
#include <stdio.h>

int is_prime(long long);

int main() {
  // find prime

  long long i,
    limit = 600851475143,
    test = 100000;

  int done = 0;

  for (i = 0; i <= test; i++) {
    if (limit == 1) return 0;
    if (is_prime(i)) {
      if (limit % i == 0) {
        printf("%lld / %lld\t= %lld\n", limit, i, limit / i);
        limit /= i;
      }
    }
  }

  return 0;
}

int is_prime(long long num) {
  long long check;

  if (num < 2) return 0;

  for (check = 2; check < num; check++) {
    if (num % check == 0) {
      return 0;
    }
  }

  return 1;
}
